File: MODULE_4/weekly_assignment/customerproduct/Ecom/src/main/java/org/northernarc/ecom/service/CustomerService.java
package org.northernarc.ecom.service;



import org.northernarc.ecom.DTO.CustomerRequestDTO;
import org.northernarc.ecom.DTO.CustomerResponseDTO;
import org.northernarc.ecom.DTO.CustomerUpdateDTO;
import org.northernarc.ecom.model.Order;

import java.util.List;

public interface CustomerService {

CustomerResponseDTO saveCustomer(CustomerRequestDTO customerDTO);

    List<CustomerResponseDTO> getAllCustomers();

    CustomerResponseDTO getCustomerById(Integer id);

    CustomerResponseDTO updateCustomer(Integer id,
                                       CustomerUpdateDTO customerDTO);

    void deleteCustomer(Integer id);

    List<Order> getCustomerOrders(Integer customerId);
}

File: MODULE_6/day5/EcomFrontEndDev/src/app/services/customer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
	CustomerRequestDTO,
	CustomerResponseDTO,
	CustomerUpdateDTO
} from '../dto/customer';
import { OrderResponseDTO } from '../dto/order';

@Injectable({
	providedIn: 'root'
})
export class CustomerService {
	private readonly baseUrl = 'http://localhost:8080/api/ecom/customer';

	constructor(private readonly http: HttpClient) {}

	getAllCustomers(): Observable<CustomerResponseDTO[]> {
		return this.http.get<CustomerResponseDTO[]>(this.baseUrl);
	}

	getCustomerById(id: number): Observable<CustomerResponseDTO> {
		return this.http.get<CustomerResponseDTO>(`${this.baseUrl}/${id}`);
	}

	addCustomer(payload: CustomerRequestDTO): Observable<CustomerResponseDTO> {
		return this.http.post<CustomerResponseDTO>(this.baseUrl, payload);
	}

	updateCustomer(id: number, payload: CustomerUpdateDTO): Observable<CustomerResponseDTO> {
		return this.http.put<CustomerResponseDTO>(`${this.baseUrl}/${id}`, payload);
	}

	deleteCustomer(id: number): Observable<void> {
		return this.http.delete<void>(`${this.baseUrl}/${id}`);
	}

	getCustomerOrders(id: number): Observable<OrderResponseDTO[]> {
		return this.http.get<OrderResponseDTO[]>(`${this.baseUrl}/${id}/orders`);
	}
}
